lib/ConstExtract/ConstExtract.cpp

```diff
@@ -0,0 +1,239 @@
+//===-------- ConstExtract.pp -- Gather Compile-Time-Known Values --------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/ConstExtract/ConstExtract.h"
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/ASTWalker.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/DiagnosticEngine.h"
+#include "swift/AST/DiagnosticsFrontend.h"
+#include "swift/AST/Evaluator.h"
+#include "swift/AST/SourceFile.h"
+#include "swift/AST/TypeCheckRequests.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/JSON.h"
+
+#include <set>
+#include <sstream>
+#include <string>
+
+namespace {
+/// A helper class to collect all nominal type declarations that conform to
+/// AppIntents protocols.
+class NominalTypeConformanceCollector : public swift::ASTWalker {
+  const std::unordered_set<std::string> &Protocols;
+  std::vector<swift::NominalTypeDecl *> &ConformanceTypeDecls;
+
+public:
+  NominalTypeConformanceCollector(
+      const std::unordered_set<std::string> &Protocols,
+      std::vector<swift::NominalTypeDecl *> &ConformanceDecls)
+      : Protocols(Protocols), ConformanceTypeDecls(ConformanceDecls) {}
+
+  bool walkToDeclPre(swift::Decl *D) override {
+    if (auto *NTD = llvm::dyn_cast<swift::NominalTypeDecl>(D))
+      for (auto &Protocol : NTD->getAllProtocols())
+        if (Protocols.count(Protocol->getName().str().str()) != 0)
+          ConformanceTypeDecls.push_back(NTD);
+    return true;
+  }
+};
+
+} // namespace
+
+namespace swift {
+
+std::unordered_set<std::string>
+parseProtocolListFromFile(StringRef protocolListFilePath,
+                          DiagnosticEngine &diags) {
+  std::unordered_set<std::string> Result;
+  // Load the input file.
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> FileBufOrErr =
+      llvm::MemoryBuffer::getFile(protocolListFilePath);
+  if (!FileBufOrErr) {
+    diags.diagnose(SourceLoc(),
+                   diag::const_extract_protocol_list_input_file_missing,
+                   protocolListFilePath);
+    return Result;
+  }
+
+  // Parse a JSON file containing a list of the format:
+  // [proto1, proto2, proto3]
+  bool ParseFailed = false;
+  {
+    StringRef Buffer = FileBufOrErr->get()->getBuffer();
+    llvm::SourceMgr SM;
+    llvm::yaml::Stream S(Buffer, SM);
```

Comment:
The input is JSON and is very basic, but this YAML parser was the simplest way I know of to parse it, since JSON is also valid YAML. If there's a more explicit/simple JSON reader, we could use that here too.

Reply:
Ah, TIL JSON is valid YAML.